scenarios/state_transitions_steady.go

```diff
@@ -0,0 +1,123 @@
+	if stateTransitionsPerSecond == 0 {
+		return fmt.Errorf("state-transitions-per-second scenario option required")
+	}
+	durationPerStateTransition := time.Second / time.Duration(stateTransitionsPerSecond)
+	s.Logger.Infof(
+		"State transitions per second is %v, which means %v between every state transition",
+		stateTransitionsPerSecond,
+		durationPerStateTransition,
+	)
+
+	// Execute initial workflow and get the transition count
+	workflowParams := kitchensink.WorkflowParams{
+		ActionSet: kitchensink.ActionSet{Actions: []*kitchensink.Action{kitchensink.NopActionExecuteActivity}},
+	}
+	workflowRun, err := s.Client.ExecuteWorkflow(
+		ctx,
+		s.NewRun(0).DefaultStartWorkflowOptions(),
+		"kitchenSink",
+		workflowParams,
+	)
+	if err != nil {
+		return fmt.Errorf("failed starting initial workflow: %w", err)
+	} else if err := workflowRun.Get(ctx, nil); err != nil {
+		return fmt.Errorf("failed executing initial workflow: %w", err)
+	}
+	resp, err := s.Client.DescribeWorkflowExecution(ctx, workflowRun.GetID(), workflowRun.GetRunID())
+	if err != nil {
+		return fmt.Errorf("failed describing workflow: %w", err)
+	}
+	stateTransitionsPerWorkflow := resp.WorkflowExecutionInfo.StateTransitionCount
+	workflowStartInterval := durationPerStateTransition * time.Duration(stateTransitionsPerWorkflow)
+	s.Logger.Infof(
+		"Simple workflow takes %v state transitions, which means we need to start a workflow every %v. "+
+			"Running for %v now...",
+		stateTransitionsPerWorkflow,
+		workflowStartInterval,
+		s.Configuration.Duration,
+	)
+
+	// Start a workflow every X interval until duration reached or there are N
+	// start failures in a row
+
+	const maxConsecutiveErrors = 5
+	errCh := make(chan error, 10000)
+	ticker := time.NewTicker(workflowStartInterval)
+	defer ticker.Stop()
+	var consecutiveErrCount int
+	iter := 1
+	for begin := time.Now(); time.Since(begin) < s.Configuration.Duration; iter++ {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case err := <-errCh:
+			if err == nil {
+				consecutiveErrCount = 0
+			} else {
+				consecutiveErrCount++
+				if consecutiveErrCount >= maxConsecutiveErrors {
+					return fmt.Errorf("got %v consecutive errors, most recent: %w", maxConsecutiveErrors, err)
+				}
+			}
+		case <-ticker.C:
+			s.Logger.Debugf("Running iteration %v", iter)
+			go func(iter int) {
+				_, err := s.Client.ExecuteWorkflow(
+					ctx,
+					s.NewRun(iter).DefaultStartWorkflowOptions(),
+					"kitchenSink",
+					workflowParams,
+				)
+				// Only send to err ch if there is room just in case
+				select {
+				case errCh <- err:
+				default:
+				}
+			}(iter)
+		}
+	}
+	// We don't want to wait to let workflows complete here. In dev cases the
+	// local server will just shutdown, in non-dev cases, the server admin can
+	// wait. But we don't want to introduce delays for those that may want to run
+	// this quickly/frequently.
+	s.Logger.Infof("Run complete, ran %v iterations", iter)
```

Comment:
I would wait.
Getting passed the loop doesn't even mean we started the workflows.
Each omes scenario should ensure completion of all of the work that it started.

Reply:
This is meant to run a certain duration steadily, not have some graceful unsteady close period.

> Each omes scenario should ensure completion of all of the work that it started.

I am not sure this should be true.

Reply:
In this case it's not even fire and forget, you may exit before starting all of the workflows if the duration is short enough and the spawn rate is fast.

> I am not sure this should be true.

I would like an operator to know by the end of a scenario run that the scenario is complete and whether it is successful or not.

Reply:
> In this case it's not even fire and forget, you may exit before starting all of the workflows if the duration is short enough and the spawn rate is fast.

I think we're ok with that

> that the scenario is complete and whether it is successful or not.

The software will complete with a successful exit code

Reply:
> The software will complete with a successful exit code

Yes, but in this case the scenario isn't complete yet (at least the way that we've defined completion in other scenarios), not sure why you want to change the definition of completion here. I'm especially concerned when someone runs a small amount of iterations to test this scenario and they may not realize that the scenario exits prematurely.

Reply:
I do think we need to do this before merging. This needs to be easy to run and orphaning workflows places operational burden on the setup that we don't want.

Reply:
The completion is when we stopped starting workflows. Heck, the workflows don't even have to succeed. I don't think we should define general completion across scenarios. Heck, some scenarios may just load test creating a bunch of namespaces and not even have workflows. That's why they're scenarios, they define what completion means to them.

Having said, that I will add a wait for max of a minute or so here to confirm workflow completion to satisfy this separate definition of completion for this scenario.

Reply:
Note, we need to be prepared when users ask why from beginning to end of the scenario the state transition count was not steady, despite the name, we need to say we have a stop period where the scenario is still alive but not creating workflows anymore.

Reply:
Added wait